src/main/java/org/apache/sling/graphql/helpers/layzloading/LazyLoadingMap.java

```diff
@@ -0,0 +1,177 @@
+        if(!(o instanceof LazyLoadingMap)) {
+            return false;
+        }
+        final LazyLoadingMap<?,?> other = (LazyLoadingMap<?,?>)o;
+
+        // Equality seems complicated to compute without this
+        computeAll();
+        return super.equals(other);
+    }
+
+    @Override
+    public int hashCode() {
+        return super.hashCode() + suppliers.hashCode();
+    }
+
+    /** Adds a Supplier that provides a lazy loaded value.
+     *  Removes existing value with the same key if it exists.
+     */
+    public Supplier<T> put(K key, Supplier<T> supplier) {
+        if(super.containsKey(key)) {
+            synchronized(this) {
+                super.remove(key);
+                return suppliers.put(key, supplier);
+            }
+        } else {
+            return suppliers.put(key, supplier);
+        }
+    }
+
+    @Override
+    public T get(Object key) {
+        return lazyCompute(key);
+    }
+
+    @SuppressWarnings("unchecked")
+    private T lazyCompute(Object key) {
+        if(key == null) {
+            return null;
+        }
+        T value = super.get(key);
+        if(value == null) {
+            synchronized(this) {
+                if(value == null) {
+                    final Supplier<T> s = suppliers.remove(key);
+                    if(s != null) {
+                        suppliersCallCount++;
+                        value = s.get();
+                        super.put((K)key, value);
+                    }
+                }
+            }
+        }
+        return value;
+    }
+
+
+    /** This indicates how many Supplier calls have been made.
+     *  Can be useful in case of doubt, as several methods need
+     *  to call computeAll().
+     */
+    public int getSuppliersCallCount() {
+        return suppliersCallCount;
+    }
+
+    @Override
+    public T remove(Object key) {
+        synchronized(this) {
+            lazyCompute(key);
+            return super.remove(key);
+        }        
+    }
+
+    @Override
+    public void clear() {
+        synchronized(this) {
+            suppliers.clear();
+            super.clear();
+        }
+    }
+
+    @Override
+    public boolean containsKey(Object key) {
+        return super.containsKey(key) || suppliers.containsKey(key);
+    }
+
+    @Override
+    public int size() {
+        return keySet().size();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return super.isEmpty() && suppliers.isEmpty();
+    }
+
+    @Override
+    public Set<K> keySet() {
+        final Set<K> result = new HashSet<>();
+        result.addAll(super.keySet());
+        result.addAll(suppliers.keySet());
+        return result;
+    }
+
+    /** Required for some methods that need all our values
+     *  Calling those methods should be avoided if possible
+     */
+    private void computeAll() {
+        log.info("computeAll called, all remaining lazy values will be evaluated now");
+        suppliers.entrySet().forEach(e -> {
```

Comment:
Does `suppliers.keySet().forEach(key -> this.get(key))` also work? Or does it cause `ConcurrentmodificationException` because `suppliers` has its elements removed during iteration? If that's the case, copying `suppliers.keySet()` before iteration should work.

On a related note: I think the current implementation needs to call `suppliers.clear()` to be correct.